File: src/main/java/com/helixd2s/minevkr/mixin/MixinMatrix4f.java
package com.helixd2s.minevkr.mixin;

import com.qouteall.immersive_portals.ducks.IEMatrix4f;
import net.minecraft.util.math.Matrix4f;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;

//mojang does not provide a method to load numbers into matrix
@Mixin(Matrix4f.class)
public class MixinMatrix4f implements IEMatrix4f {
    @Shadow public float a00;
    @Shadow public float a01;
    @Shadow public float a02;
    @Shadow public float a03;
    @Shadow public float a10;
    @Shadow public float a11;
    @Shadow public float a12;
    @Shadow public float a13;
    @Shadow public float a20;
    @Shadow public float a21;
    @Shadow public float a22;
    @Shadow public float a23;
    @Shadow public float a30;
    @Shadow public float a31;
    @Shadow public float a32;
    @Shadow public float a33;
    
    @Override
    public void loadFromArray(float[] arr) {
        a00 = arr[0];
        a01 = arr[1];
        a02 = arr[2];
        a03 = arr[3];
        a10 = arr[4];
        a11 = arr[5];
        a12 = arr[6];
        a13 = arr[7];
        a20 = arr[8];
        a21 = arr[9];
        a22 = arr[10];
        a23 = arr[11];
        a30 = arr[12];
        a31 = arr[13];
        a32 = arr[14];
        a33 = arr[15];
    }
}

File: src/main/kotlin/com/helixd2s/minevkr/mixin/MixinMatrix4f.kt
package com.helixd2s.minevkr.mixin


import com.helixd2s.minevkr.ducks.IEMatrix4f
import net.minecraft.util.math.Matrix4f
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow

//mojang does not provide a method to load numbers into matrix
@Mixin(Matrix4f::class)
class MixinMatrix4f : IEMatrix4f {
    @Shadow var a00 = 0f
    @Shadow var a01 = 0f
    @Shadow var a02 = 0f
    @Shadow var a03 = 0f
    @Shadow var a10 = 0f
    @Shadow var a11 = 0f
    @Shadow var a12 = 0f
    @Shadow var a13 = 0f
    @Shadow var a20 = 0f
    @Shadow var a21 = 0f
    @Shadow var a22 = 0f
    @Shadow var a23 = 0f
    @Shadow var a30 = 0f
    @Shadow var a31 = 0f
    @Shadow var a32 = 0f
    @Shadow var a33 = 0f

    // TODO: Dynamic Array Size (12x or 16x)
    override fun loadFromArray(arr: FloatArray?) {
        a00 = arr?.get(0) ?: 1.0F
        a01 = arr?.get(1) ?: 0.0F
        a02 = arr?.get(2) ?: 0.0F
        a03 = arr?.get(3) ?: 0.0F
        a10 = arr?.get(4) ?: 0.0F
        a11 = arr?.get(5) ?: 1.0f
        a12 = arr?.get(6) ?: 0.0F
        a13 = arr?.get(7) ?: 0.0F
        a20 = arr?.get(8) ?: 0.0F
        a21 = arr?.get(9) ?: 0.0F
        a22 = arr?.get(10) ?: 1.0F
        a23 = arr?.get(11) ?: 0.0F
        a30 = arr?.get(12) ?: 0.0F
        a31 = arr?.get(13) ?: 0.0F
        a32 = arr?.get(14) ?: 0.0F
        a33 = arr?.get(15) ?: 1.0F
    }

    override fun toArray(): FloatArray {
        return floatArrayOf(
                a00, a01, a02, a03,
                a10, a11, a12, a13,
                a20, a21, a22, a23,
                a30, a31, a32, a33
        )
    }
}
